RHOARDOC-1550: Refactor Deploying an application to OpenShift

theashiot opened the pull request:
Initial PR only for seeing rendered doc for vertx-runtime.

If the output is ok, I will create the proper assemblies and modules (with correct filenames, ids etc).

There will be typos, please only review the structure:

Developing and deploying an application
* Developing an application
* Deploying on Openshift
** Preparing to deploy
** Deploy using fabric8
* Deploying standalone RHEL
** Preparing to deploy
** Deploying using java

tradej commented:
ok to test

tradej commented:
Seems reasonable to me. @liborfuka, can you please verify this structure and rough sketch of a solution looks good to you (only the Vert.x runtime at the moment)? I know it was you who requested this change during the Spring Boot 2.1 release. Thanks.

liborfuka commented:
Please, send me a link to doc in html format.
Thank you.

akoniar commented:
@liborfuka link for spring-boot doc: https://deploy-preview-1264--launcher-documentation.netlify.com/docs/spring-boot-runtime.html

liborfuka commented:
@akoniar This PR is for VertX

akoniar commented:
@liborfuka Thanks, I don't read the above comment. I'll look at it.

akoniar commented:
@liborfuka @tradej the structure looks good to me.

liborfuka commented:
looks good to me.

tradej commented:
Thank you, gentlemen; @theashiot, feel free to proceed with the changes.

tradej commented on docs/topics/proc_deploying-an-application-to-openshift.adoc:
```diff
@@ -7,19 +7,25 @@
-[id='deploying-an-application-to-openshift_{context}']
-= Deploying an application to OpenShift
+[id='deploying-an-application-to-openshift-using-fabric8-launcher_{context}']
+= Deploying {runtime} application to OpenShift using {name-launcher} 
```
I believe this should say _using Fabric8 Maven Plugin_ instead of the Launcher.

tradej commented on docs/topics/assembly_deploying-an-application-to-openshift.adoc:
```diff
@@ -0,0 +1,13 @@
+[id='assembly-deploying-an-application-to-openshift_{context}']
```
The `assembly-` prefix should not be in the ID.

tradej commented:
@theashiot, thank you for the content. Once you believe it's ready for review, please indicate so and remove the draft designation, we'll run the round of SME and peer reviews; this is a lot of changes, I expect to have quite a few back and forths, thanks.

theashiot commented:
Thanks @tradej for spotting the typo and the suggestions. I think I'll do another round of review to see if there are any small things I missed. Once I do that, I'll remove the WIP status.

best,

ashwin

theashiot commented:
Looks like there are merge conflicts. The two files that are conflicting are used by Thorntail doc. According to https://launcher.fabric8.io/docs/contrib-guide.html#git-workflow-for-making-changes_contributing we are not supposed to edit files in docs/topics/thorntail directory. But the files in conflict are in docs/topics/ directory. Should these files be reverted to original @tradej ?

tradej commented:
@theashiot Not reverted, you need to rebase the branch (`git rebase upstream/master`) and resolve the conflicts manually, i. e. make sure that any changes made to `master` in the meantime are also incorporated in your PR. I recommend you squash all the commits in your branch before you start resolving the conflicts as you'll only have to resolve the conflict once. Have you done this before?

theashiot commented:
Hi @tradej,

I resolved the conflicts and also made some changes and now the PR is ready to be reviewed.

As per your previous comment, I have updated filenames and ids to match each other.

A word of caution - since I squashed commits, I had to force-push the latest commit. 

Can you please review the PR?

best,
ashwin

theashiot commented on docs/nodejs-runtime/master.adoc:
```diff
@@ -122,15 +122,16 @@ ifdef::launcher-docs-only[]
+include::topics/assembly_developing-and-deploying-a-nodejs-application.adoc[leveloffset=+1]
 
-include::topics/con_nodejs-runtime-api.adoc[leveloffset=+2]
+//include::topics/con_nodejs-runtime-api.adoc[leveloffset=+2]
```
Had commented out while working. Forgot to remove, thanks for spotting! In fact, the comment above "== Developing an application for the {Node} runtime" is also a remnant of work in progress. Changing it to a more meaningful comment: "DEVELOPING AN APPLICATION".

theashiot commented on docs/topics/assembly_deploying-runtime-application-to-openshift.adoc:
```diff
@@ -0,0 +1,13 @@
+= Deploying {runtime} application to OpenShift
+
+ifdef::built-for-vertx[]
+To deploy an {runtime} application to OpenShift, you must configure the application's `pom.xml` file. You can then deploy the application using Fabric8 Maven launcher.
```
Agreed. Nice solution. 

- Changed to "your" {runtime} application.
- Rewrote the sentence to avoid Saxon genitive: 
 "To deploy your {runtime} application to OpenShift, configure the pom.xml file in your application and then deploy using Fabric8 Maven plugin."
- Changed *Fabric8 Maven launcher* to *Fabric8 Maven plugin*.

theashiot commented on docs/topics/assembly_deploying-runtime-application-to-standalone-red-hat-enterprise-linux.adoc:
```diff
@@ -0,0 +1,13 @@
+= Deploying {runtime} application to standalone Red Hat Enterprise Linux
+
+ifdef::built-for-vertx[]
+To deploy an {runtime} application to standalone Red{nbsp}Hat Enterprise Linux, you must first package the application. You can then deploy the application using `java -jar` command. 
```
Done.

theashiot commented on docs/topics/assembly_developing-and-deploying-a-nodejs-application.adoc:
```diff
@@ -0,0 +1,8 @@
+[id='developing-and-deploying-a-nodejs-application_{context}']
+= Developing and deploying a {Node} application
+
+The {Node} runtime provides the core {Node} API which is documented in the link:https://nodejs.org/api/[{Node} API documentation^]. 
```
Removed from the assembly, added as additional information in Developing nodejs application. Does that make sense?

theashiot commented:
Hi @liborfuka, can you please review the changes made to vertx, springboot and nodejs runtimes and let me know if these changes are ok to apply to thorntail runtime as well? On approval, I'll create an PR for thorntail repo.

best,

ashwin

liborfuka commented:
I can't verify vertx, nodejs and thorntail.
SpringBoot 1.5.x is not certified with OpenJDK11, please remove all OpenJDK11 from 1.5.x runtime guide.

tradej commented:
@liborfuka Thanks for the comment. Are there any other things that need changing? I would prefer if you can provide as much feedback as possible in one go to save time, thanks.

liborfuka commented:
Chapter 5 is ok in SpringBoot 2.1.x guide.  
SpringBoot 1.5.x guide needs fix - remove OpenJDK11.

liborfuka commented:
SpringBoot Guides 2.1.x and 1.5.x verified.

theashiot commented:
> In addition to my in-line comments, I can't see any changes to the Thorntail sources. @theashiot, have you made any changes there?

Changes to Thorntail files native to this repository made. Issuing PR to thornatil repo for changes required in "Creating a thorntail application from scratch".

tradej commented:
I've got an ack from David Simansky over IRC.

theashiot commented:
@tradej draft PR submitted to thorntail repo: https://github.com/thorntail/thorntail/pull/1281

lholmquist commented on docs/topics/proc_deploying-a-nodejs-application-to-openshift.adoc:
```diff
@@ -0,0 +1,52 @@
+$ npm run openshift
+----
+
+. Expose the application in OpenShift using link:https://docs.openshift.com/online/dev_guide/routes.html[a route].
```
Is it possible to remove this whole section related to the routes and instead add the `--expose` flag to the "openshift" npm script?

I would also like to get away from using the `.nodeshift` directory if possible

lholmquist commented on docs/topics/proc_preparing-nodejs-application-for-openshift-deployment.adoc:
```diff
@@ -0,0 +1,70 @@
++
+[source,bash,options="nowrap",subs="attributes+"]
+----
+$ npm install nodeshift --save
```
we usually use --save-dev here since it isn't needed to actually run the application,  just deploy it

lholmquist commented on docs/topics/proc_preparing-nodejs-application-for-openshift-deployment.adoc:
```diff
@@ -0,0 +1,70 @@
+  "main": "app.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "openshift": "nodeshift --strictSSL=false --metadata.out=deployment-metadata.json --build.forcePull=true {version-node-nodeshift}",
```
you can probably get rid of most of those flag and it could be `"openshift":"nodeshift --expose {version-node-nodeshift}" `

I've added the `--expose` flag here which creates a route for you.  I mentioned in a previous comment that maybe we do the route thing here.

I would also like to get away from using that `.nodeshift` directory if possible

lholmquist commented on docs/topics/proc_deploying-a-nodejs-application-to-openshift.adoc:
```diff
@@ -0,0 +1,52 @@
+.Example `.nodeshift/route.yaml`
+[source,yaml,options="nowrap",subs="attributes+"]
+----
+apiVersion: v1
```
If we do need to keep this section,  please remove the apiVersion here since the latest version of nodeshift will fail since the api version has changed to `route.openshift.io/v1`

tradej commented:
Filed a follow-up [JIRA](https://issues.jboss.org/browse/RHOARDOC-1640) for @lholmquist's comments.
